File: diff/comparison/patience_matcher.go
package comparison

import (
	"strings"
	"unicode"

	"github.com/byron1st/file-diff/diff/patience"
)

// PatienceMatcher implements LineMatcher using the Patience Diff algorithm.
// It anchors on unique common lines via LIS, then falls back to Myers
// for regions without unique lines.
type PatienceMatcher struct{}

func (m *PatienceMatcher) Match(left, right []string, policy ComparisonPolicy) FairDiffIterable {
	normLeft := normalizeLines(left, policy)
	normRight := normalizeLines(right, policy)

	matches := patience.Diff(normLeft, normRight, func(l, r []string) []patience.Anchor {
		return myersFallback(l, r)
	})

	builder := NewChangeBuilder(len(left), len(right))
	for _, match := range matches {
		builder.MarkEqual(match.LeftIdx, match.RightIdx)
	}
	return Fair(builder.Finish())
}

// normalizeLines returns line contents normalized by policy for comparison.
func normalizeLines(lines []string, policy ComparisonPolicy) []string {
	if policy == PolicyDefault {
		return lines
	}
	result := make([]string, len(lines))
	for i, l := range lines {
		result[i] = normalizedContent(l, policy)
	}
	return result
}

func normalizedContent(s string, policy ComparisonPolicy) string {
	switch policy {
	case PolicyTrimWhitespaces:
		return strings.TrimSpace(s)
	case PolicyIgnoreWhitespaces:
		var b strings.Builder
		for _, c := range s {
			if !unicode.IsSpace(c) {
				b.WriteRune(c)
			}
		}
		return b.String()
	default:
		return s
	}
}

// myersFallback uses CompareLines (Myers) to produce matched pairs.
func myersFallback(left, right []string) []patience.Anchor {
	if len(left) == 0 || len(right) == 0 {
		return nil
	}
	result := CompareLines(left, right, PolicyDefault)
	var matches []patience.Anchor
	for _, u := range result.Unchanged() {
		count := u.End1 - u.Start1
		for i := range count {
			matches = append(matches, patience.Anchor{LeftIdx: u.Start1 + i, RightIdx: u.Start2 + i})
		}
	}
	return matches
}

File: swift/Sources/FileDiff/PatienceMatcher.swift
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Swift (via the Go port in this repository).

/// Line matcher backed by the Patience Diff algorithm.
///
/// Anchors on unique common lines via longest increasing subsequence, then
/// delegates regions without unique anchors to a Myers-based fallback.
public struct PatienceMatcher: LineMatcher {
    public init() {}

    public func match(_ left: [String], _ right: [String], policy: ComparisonPolicy) -> FairDiffIterable {
        let normLeft = normalizeLines(left, policy: policy)
        let normRight = normalizeLines(right, policy: policy)

        let matches = patienceDiff(normLeft, normRight) { subLeft, subRight in
            myersFallback(subLeft, subRight)
        }

        let builder = ChangeBuilder(length1: left.count, length2: right.count)
        for match in matches {
            builder.markEqual(match.leftIndex, match.rightIndex)
        }
        return fair(builder.finish())
    }
}

func normalizeLines(_ lines: [String], policy: ComparisonPolicy) -> [String] {
    if policy == .default {
        return lines
    }
    return lines.map { normalizedContent($0, policy: policy) }
}

private func normalizedContent(_ line: String, policy: ComparisonPolicy) -> String {
    switch policy {
    case .trimWhitespaces:
        line.trimmingCharacters(in: .whitespacesAndNewlines)
    case .ignoreWhitespaces:
        String(line.unicodeScalars.filter { !CharacterUtils.isWhiteSpaceCodePoint($0) })
    default:
        line
    }
}

func myersFallback(_ left: [String], _ right: [String]) -> [PatienceAnchor] {
    if left.isEmpty || right.isEmpty {
        return []
    }
    let result = compareLines(left, right, policy: .default)
    var matches: [PatienceAnchor] = []
    for unchanged in result.unchanged {
        let count = unchanged.left.count
        for i in 0..<count {
            matches.append(PatienceAnchor(
                leftIndex: unchanged.left.lowerBound + i,
                rightIndex: unchanged.right.lowerBound + i
            ))
        }
    }
    return matches
}
